Features branch

AhmadKHamdan opened the pull request.

AhmadKHamdan committed:
Add initial models and project files for Restaurant Reservation system

AhmadKHamdan committed:
Refactor RestaurantReservationDbContext to include DbSet properties and configure database options

AhmadKHamdan committed:
Add initial migration and fixed DbContext to include all existing models

AhmadKHamdan committed:
Fix foreign key relationship in OrderItems for MenuItem; Insert 5 rows for each table

AhmadKHamdan committed:
Refactor project structure: move models and services to separate projects; implement UserMenu for main application flow

AhmadKHamdan committed:
Change service classes from internal to public access modifier

AhmadKHamdan committed:
Refactor project structure: update namespaces and implement RestaurantService for managing restaurant data

AhmadKHamdan committed:
Refactor service and repository structure: implement AllRepositories and AllServices for improved organization and dependency management

AhmadKHamdan committed:
Refactor program structure: implement dependency injection for services and remove AllRepositories and AllServices classes

AhmadKHamdan committed:
Implement ViewAll methods for services and override ToString in models for better output representation

AhmadKHamdan committed:
Add AddDataMenu method and validation for restaurant input; refactor Restaurant model properties

AhmadKHamdan committed:
Implement customer validation and add customer management functionality; refactor related services and repositories

AhmadKHamdan committed:
Add validation logic for various entities and update validation messages; implement Employee, MenuItem, Order, Reservation, and Table validators

AhmadKHamdan committed:
Add repository methods for adding entities and retrieving by ID; implement Add and GetById methods in Customer, Employee, MenuItem, Order, Reservation, and Table repositories

AhmadKHamdan committed:
Refactor input handling by introducing InputHelper class; streamline input validation across services and enhance menu options in UserMenu

AhmadKHamdan committed:
Implement delete functionality for various entities; add Delete methods in services and repositories for Customer, Employee, MenuItem, Order, Reservation, Table, and OrderItem

AhmadKHamdan committed:
Implement update functionality for managing various entities; add Update methods in services for Customer, Employee, MenuItem, Order, Reservation, Restaurant, and Table

AhmadKHamdan committed:
Add second menu and list managers functionality;

AhmadKHamdan committed:
Implement ListReservationsByCustomer method and update console prompts for better readability

AhmadKHamdan committed:
Implement order and menu item listing functionality; add methods to ReservationService and repositories for better order management

AhmadKHamdan committed:
Add CalculateAverageOrderAmountByEmployee method in OrderService; Refactor input handling and improve average order calculation; replace Console.ReadKey with InputHelper.Continue for consistency

AhmadKHamdan committed:
- Added SQL scripts to create database views for reservations and employee details.
- Created ViewService to display the database views.
- Introduced new DTOs: EmployeeDetailsDTO and ReservationDetailsDTO for handling the views data.
- Added ViewRepository for accessing reservation and employee details from the database views.
- Refactored the services for better folder management

AhmadKHamdan committed:
Implement CalcTotalRevenue function; Add FunctionRepository and FunctionService

AhmadKHamdan committed:
Add stored procedure for listing customers by minimum party size; refactor services and repositories for better organization

AhmadKHamdan committed:
Refactor menu structure: replace UserMenu with MainMenu and add dedicated menus for adding, deleting, managing data, and querying; update service registrations accordingly.

oalataiwi commented on RestaurantReservation.ConsoleApp/Menus/AddDataMenu.cs:
```diff
@@ -0,0 +1,96 @@
+using RestaurantReservation.Services.Helpers;
```
I don’t think these menu services are necessary.
You don’t need to display all options to the user.

The expected deliverables for this task are:

1. The database layer, including migrations and DbContext
2. The repositories and services that perform the required operations

oalataiwi commented on RestaurantReservation.ConsoleApp/Program.cs:
```diff
@@ -0,0 +1,91 @@
+            services.AddDbContext<RestaurantReservationDbContext>();
+
+            // Register Repositories 
+            services.AddScoped<RestaurantRepository>();
```
this is not needed as the first comment i stated
u do not need to inject these services for now

oalataiwi replied:
it is up to u

oalataiwi commented on RestaurantReservation.Core/Models/Customer.cs:
```diff
@@ -0,0 +1,18 @@
+﻿namespace RestaurantReservation.Core.Models
```
the layer that contains the database models is more accurately referred to as the database layer or infrastructure layer, not the core. in software architecture, the core is generally understood as the domain layer

oalataiwi commented on RestaurantReservation.Db/Repositories/FunctionRepository.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace RestaurantReservation.Db.Repositories
+{
+    public class FunctionRepository
```
we don’t include technical details like this in repo or service names.
the GetRestaurantRevenue method should just gto restaurantRepo

oalataiwi commented on RestaurantReservation.Db/Repositories/StoredProcedureRepository.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace RestaurantReservation.Db.Repositories
+{
+    public class StoredProcedureRepository
```
the same applies here, please check which repo is most appropriate to place the FindCustomersByPartySize method in.

oalataiwi commented on RestaurantReservation.Db/Repositories/ViewRepository.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace RestaurantReservation.Db.Repositories
+{
+    public class ViewRepository
```
the same applies here, please check which repo is most appropriate to place methods in this repo

oalataiwi commented on RestaurantReservation.Services/Helpers/InputHelper.cs:
```diff
@@ -0,0 +1,199 @@
+
+namespace RestaurantReservation.Services.Helpers
+{
+    public static class InputHelper
```
not needed, we do not need to handle anything from the user, just we need services and repos so we can continue from there in the next task

oalataiwi commented on SSMS/Special Queries.sql:
```diff
@@ -0,0 +1,22 @@
+-- Create View 1: 
```
this could be handled through a migration, why did you implement it in this script?

oalataiwi commented on RestaurantReservation.Services/SpecialServices/ViewService.cs:
```diff
@@ -0,0 +1,66 @@
+{
+    public class ViewService
+    {
+        private readonly ViewRepository _viewRepository;
```
all these should be  removed
ViewService
StoredProcedureService
FunctionService

oalataiwi commented on RestaurantReservation.Db/Repositories/CustomerRepository.cs:
```diff
@@ -0,0 +1,47 @@
+            _context = context;
+        }
+
+        public List<Customer> GetAll()
+        {
+            return _context.Customers.ToList();
+        }
```
all db calls must be async

 public Task<List<Customer>> GetAllAsync()
        {
            return  await _context.Customers.ToListAsync();
        }

oalataiwi commented on RestaurantReservation.Db/Repositories/CustomerRepository.cs:
```diff
@@ -0,0 +1,47 @@
+            return _context.Customers.ToList();
+        }
+
+        public void Add(Customer customer)
```
plz make all db ops async . this ensures non-blocking i/o, better scalability under load. consequently, the service methods that depend on these db calls should also be updated to async to propagate the async flow properly.

oalataiwi commented on RestaurantReservation.Services/MainServices/CustomerService.cs:
```diff
@@ -0,0 +1,125 @@
+            _customerRepo = customerRepo;
+        }
+
+        public void ViewAll()
```
this should be 
`public Task<List<Customer>> ViewAllAsync() `

we wont print anything to console, we should do simple return,

oalataiwi commented on RestaurantReservation.Services/MainServices/CustomerService.cs:
```diff
@@ -0,0 +1,125 @@
+
+        public void ViewAll()
+        {
+            if (IsEmpty())
```
no need to check if the table is empty, just return whatever is in the db (it could simply be an empty list).

id u need to check if it is empty, u can check the returned list from repo

oalataiwi commented on RestaurantReservation.Services/MainServices/CustomerService.cs:
```diff
@@ -0,0 +1,125 @@
+            Console.WriteLine();
+            try
+            {
+                var firstName = InputHelper.GetValidInput(ValidationMessages.EnterFirstName, CustomerValidator.ValidateFirstName);
+                var lastName = InputHelper.GetValidInput(ValidationMessages.EnterLastName, CustomerValidator.ValidateLastName);
+                var email = InputHelper.GetValidInput(ValidationMessages.EnterEmail, CustomerValidator.ValidateEmail);
+                var phoneNumber = InputHelper.GetValidInput(ValidationMessages.EnterPhone, CustomerValidator.ValidatePhoneNumber);
```
we simply pass the object we recieve in the params, validate it, then add it via repo

oalataiwi commented on RestaurantReservation.Services/MainServices/CustomerService.cs:
```diff
@@ -0,0 +1,125 @@
+            }
+        }
+
+        public void Delete()
+        {
+            Console.WriteLine();
+            try
+            {
+                var customerId = InputHelper.GetValidCustomerId(_customerRepo);
+                var customer = _customerRepo.GetById(customerId);
+
+                if (customer == null)
+                {
+                    Console.WriteLine($"Customer not found.");
+                }
+                else
+                {
+                    _customerRepo.Delete(customer);
+                    Console.WriteLine($"Customer with ID {customerId} deleted successfully.");
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Error deleting customer: {ex.Message}");
+            }
+        }
```
```
public async Task DeleteAsync(int customerId)
{
    var customer = await _customerRepo.GetByIdAsync(customerId);

        if (customer == null)
        {
           //throw exception 404
        }

        await _customerRepo.DeleteAsync(customer); 
}
```

oalataiwi commented on RestaurantReservation.Services/MainServices/CustomerService.cs:
```diff
@@ -0,0 +1,125 @@
+            }
+        }
+
+        private bool IsEmpty()
```
i don’t think this is necessary. and if it really is, there’s no need to create a separate service method — just call repo.isEmpty directly from within the service.

AhmadKHamdan committed:
Refactor: Remove old menu classes and input helper

- Deleted ManageDataMenu, QueriesMenu, and ViewDataMenu classes to streamline the console application.
- Removed InputHelper class as its functionality is no longer needed.
- Updated Program.cs to remove references to deleted menus.
- Moved model classes from Core to Db.Models namespace for better organization.
- Updated service classes to reference new Db.Models instead of Core.Models.
- Adjusted CustomerService and removed all console related work.

AhmadKHamdan committed:
Refactor services to follow best practice for API-ready methods

AhmadKHamdan committed:
Refactor repositories and services: remove special repos and services and implement them in their respective correct places

AhmadKHamdan committed:
Refactor: Remove manual views creation, implementing through migrations

AhmadKHamdan committed:
Refactor repositories to use asynchronous methods for retrieving all entities

AhmadKHamdan committed:
Refactor repositories to use asynchronous methods for adding entities

AhmadKHamdan committed:
Refactor repositories to use asynchronous methods for GetById operations

AhmadKHamdan committed:
Refactor repositories to use asynchronous methods for Update operations

AhmadKHamdan committed:
Refactor repositories to rename Update methods to UpdateAsync for consistency

AhmadKHamdan committed:
Refactor repositories to use asynchronous methods for Delete operations

AhmadKHamdan committed:
Refactor repositories to use asynchronous methods for various operations

AhmadKHamdan committed:
Refactor services: Delete the main services directory, move all services directly under services project

AhmadKHamdan committed:
Refactor services and repositories to use asynchronous methods for all operations

AhmadKHamdan committed:
Refactor services and validation: Update methods to use asynchronous patterns and improve name validation logic

AhmadKHamdan committed:
Refactor repositories and services: Implement repository and service interfaces for better abstraction and dependency injection

AhmadKHamdan committed:
Add initial minimal API project structure with configuration and service registration; Remove consoleApp project

AhmadKHamdan committed:
Implement table management endpoints and DTOs; Fixed an issue with services expecting concrete repositories implementation

AhmadKHamdan committed:
Add employee management endpoints and DTOs; Enhance employee and table repository methods to include restaurant details

AhmadKHamdan committed:
Refactor DTOs: Move employee and table DTOs to their respective directories and remove old implementations

AhmadKHamdan committed:
Add restaurant management endpoints and DTOs

AhmadKHamdan committed:
Add customer management endpoints, DTOs, and update customer service methods

AhmadKHamdan committed:
Add menu item management endpoints, DTOs, and service methods

AhmadKHamdan committed:
Add order management endpoints, DTOs, and update order service and repo; Enhance functionality in all services

AhmadKHamdan committed:
Add order item and reservation management endpoints, DTOs, and update repositories

AhmadKHamdan committed:
Add endpoint to retrieve all employees with the position of 'Manager' and update GetManagersAsync to include restaurant details

AhmadKHamdan committed:
Refactor DTOs to remove required properties for IDs and update service methods to handle nullable relationships; Add OrderWithItemsDTO for enhanced order retrieval

AhmadKHamdan committed:
Add JWT authentication support with token generation and validation; Update appsettings for JWT configuration

AhmadKHamdan committed:
Add authorization requirement to all API endpoints and implement JWT authentication setup

AhmadKHamdan committed:
Add authorization requirement to table endpoints and remove unnecessary properties from service models

AhmadKHamdan committed:
Fix phone number validation to return error message instead of logging to console

AhmadKHamdan committed:
Refactor the project for a better layered architecture; Removed service layer and implemented it in core layer; Added shared layer for DTOs

AhmadKHamdan committed:
Refactor DTO namespaces and add validation classes

- Updated namespaces for Customer, Employee, MenuItem, Order, OrderItem, Reservation, Restaurant, and Table DTOs to follow the shared structure.
- Implemented CreateCustomerValidator, UpdateCustomerValidator, PartySizeValidator for customer-related validations.
- Added CreateEmployeeValidator and UpdateEmployeeValidator for employee-related validations.
- Created CreateMenuItemValidator and UpdateMenuItemValidator for menu item validations.
- Developed CreateOrderValidator and UpdateOrderValidator for order validations.
- Introduced CreateReservationValidator and UpdateReservationValidator for reservation validations.
- Added CreateRestaurantValidator and UpdateRestaurantValidator for restaurant validations.
- Implemented CreateTableValidator and UpdateTableValidator for table validations.
- Introduced PartySizeDTO to encapsulate party size information.

AhmadKHamdan committed:
Enhance error handling and validation in API endpoints; return structured error responses for not found and bad request scenarios

AhmadKHamdan committed:
Implement fluent validation for Orders, OrderItems and Reservations

AhmadKHamdan committed:
Implement fluent validation for Restaurants and Tables

AhmadKHamdan committed:
Added detailed API documentation

AhmadKHamdan committed:
Refactor GenerateToken to accept TokenRequest object; Enhanced exception handling by implementing ExceptionHandler; Added NotFoundException for handling not found errors; Enhanced customers and employees endpoints clarity and cleaniness by moving all data logic to their services

AhmadKHamdan committed:
Enhanced menu items, order items and orders endpoints clarity and cleaniness by moving all data logic to their services

AhmadKHamdan committed:
Enhanced reservations, restaurants and tables endpoints clarity and cleaniness by moving all data logic to their services

AhmadKHamdan committed:
Moved employee positions enum to a separate shared file; Moved validation messages to a shared file

AhmadKHamdan committed:
Moved custom regex to a separate shared folder for reuse

AhmadKHamdan committed:
Removed constant magic numbers to a separate file

AhmadKHamdan committed:
Added automapping to customers service

AhmadKHamdan committed:
Added auto mappers for all entities to map entities to DTOs

AhmadKHamdan committed:
Merged createOrderValidator and updateOrderValidator into orderValidator

AhmadKHamdan committed:
Merge pull request #2 from Ahmad-K-Hamdan/APIs-Integration

APIs integration
